### PR TITLE
Add attn_logit_softcapping config, plumbed to TE and local attention

### PR DESCRIPTION
**Pull request by nvegesna-netizen**
Exposes tanh logit soft-capping as a first-class config field so models with this attention variant (e.g. Gemma2) can run on TransformerEngine's FlashAttention backends instead of an unfused fallback. Note that softcapping disqualifies cuDNN `FusedAttention` in TE, so the target here is FA2/FA3, not the fused backend.

Companion PRs:
- TransformerEngine: NVIDIA/TransformerEngine#3391 — adds the underlying `softcap` kwarg this maps to
- Megatron-Bridge: NVIDIA-NeMo/Megatron-Bridge#5628 — consumes this field for an opt-in Gemma2 attention module

### What changed
- `TransformerConfig.attn_logit_softcapping: Optional[float] = None` — `None` disables softcapping, default behavior unchanged. Rejected at config time if not positive and finite: `0.0` means "disabled" in TransformerEngine but collapses every logit to zero on the local path, a negative cap is applied as its absolute value locally while FlashAttention ignores it, and a non-finite cap yields NaN. Failing in `__post_init__` keeps the two conventions from quietly meaning different things.
- `TEDotProductAttention` maps this to TE's `softcap` kwarg via a signature capability probe on the installed TE build, rather than a version gate — this lets it work against TE builds that add `softcap` support ahead of a version bump, and asserts clearly if the field is set but the installed TE build doesn't support it.
- Local (no-TE) `DotProductAttention` applies the equivalent cap pre-softmax. The cap is resolved in `__init__` next to `coeff`, because `apply_query_key_layer_scaling` divides `softmax_scale` by `layer_number` and defers the matching multiply to `scale_mask_softmax`; without dividing the cap the same way, the softmax would see an effective cap of `cap * layer_number`, growing with depth.
- The direct flash-attention inference paths now honour the cap too. They bypass `self.core_attention` and call the kernels directly, so the value is plumbed out through a `_get_inference_softcap` helper mirroring the existing `_get_inference_softmax_offset`. Where a kernel cannot carry the cap it now fails rather than dropping it: the FA3 wrapper filters kwargs against the runtime signature and so asserts, and FlashMLA's decode kernel has no softcap support and refuses the combination rather than capping in prefill only. The kwarg is passed only when a cap is configured, so runs without softcapping keep their existing call signatures.

### Validation
Exercised end-to-end via the companion TransformerEngine and Megatron-Bridge changes above — multi-step training runs confirmed numerically consistent with the prior unfused path. Those runs predate the FA3-default change in TE#3391, so on Hopper they exercised FA2; FA3 is now the default softcap backend on sm90 when a softcap-capable build is installed.

The three fixes above are reasoned from the code rather than measured: the magnitude bug and the `0.0` case both require configurations (`apply_query_key_layer_scaling=True`, or a cap of exactly `0.0`) that an end-to-end Gemma2 run does not exercise. This PR ships no unit tests.

### CODE REVIEW

**Comment by nvegesna-netizen** [comment]
@NVIDIA/mcore-oncall can you review?

**Review by Phlip79 — CHANGES_REQUESTED** [review]
Blocking until https://github.com/NVIDIA/TransformerEngine/pull/3391 is merged

**Review by Connor-XY — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

File: megatron/core/transformer/attention.py [hunk 1]
```diff
@@ -810,6 +816,17 @@ def flash_decode(
             out = flash_attn_with_kvcache(**kv_kwargs)
         return out
 
+    def _get_inference_softcap(self) -> Optional[float]:
```
Comment (jaredcasper) [inline megatron/core/transformer/attention.py]: Why is this a function?

Comment (nvegesna-netizen) [reply megatron/core/transformer/attention.py]: No good reason, and it is gone in `3f98a5fd7`.

It only returned `self.config.attn_logit_softcapping`. I had written it to mirror `_get_inference_softmax_offset`, but that one earns its place: it branches on `softmax_type`, falls back through `getattr`, asserts, and lazily caches a zero tensor. Mine was a pass-through wrapped in a docstring, and this file reads `self.config.<field>` directly everywhere else, so the indirection just added a layer to step through.

Now inlined at all six sites, which is a net eight lines shorter. I kept one comment at the `flash_decode` site for the part that is not inferable from the code, that these paths bypass `self.core_attention` and so need the cap passed again.

**Review by jaredcasper — APPROVED** [review]

**Review by guihong-nv — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]
